### ISSUE TITLE
Unable to build docker image- Win10

### ISSUE THREAD

**Issue by saikumaru** [chatwoot/chatwoot#3233]
**Describe the bug**
`docker-compose build` fails on Win-10 

**To Reproduce**

https://www.chatwoot.com/docs/contributing-guide/environment-setup/docker
1 Clone chatwoot on win-10
2 Follow instructions for docker build
3 docker-compose build
4 See error

**Expected behavior**
Docker build must be successfull.

**Screenshots**

**Browser logs**

Share the browser logs to debug the issue further

**Server logs**

```
docker-compose build --no-cache
time="2021-10-16T22:13:59+05:30" level=warning msg="The \"REDIS_PASSWORD\" variable is not set. Defaulting to a blank string."
[+] Building 2.4s (10/10) FINISHED
 => [chatwoot:development internal] load build definition from Dockerfile                                          0.0s
 => => transferring dockerfile: 2.77kB                                                                             0.0s
 => [chatwoot-webpack:development internal] load build definition from webpack.Dockerfile                          0.0s
 => => transferring dockerfile: 163B                                                                               0.0s
 => [chatwoot-rails:development internal] load build definition from Dockerfile                                    0.0s
 => => transferring dockerfile: 32B                                                                                0.0s
 => [chatwoot:development internal] load .dockerignore                                                             0.0s
 => => transferring context: 35B                                                                                   0.0s
 => [chatwoot-rails:development internal] load .dockerignore                                                       0.0s
 => => transferring context: 35B                                                                                   0.0s
 => [chatwoot-webpack:development internal] load .dockerignore                                                     0.0s
 => => transferring context: 35B                                                                                   0.0s
 => CANCELED [chatwoot-rails:development internal] load metadata for docker.io/library/ruby:3.0.2-alpine           2.1s
 => ERROR [chatwoot-webpack:development internal] load metadata for docker.io/library/chatwoot:development         2.0s
 => [auth] library/ruby:pull token for registry-1.docker.io                                                        0.0s
 => [auth] library/chatwoot:pull token for registry-1.docker.io                                                    0.0s
------
 > [chatwoot-webpack:development internal] load metadata for docker.io/library/chatwoot:development:
------
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```

**Environment**

Development environment/ ASUS Win-10
Docker setup

**Desktop (please complete the following information):**
 - Win-10

**Smartphone (please complete the following information):**
NA

**Additional context**

```
Client:
 Cloud integration: 1.0.17
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:58:50 2021
 OS/Arch:           windows/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:52:31 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

```

Also tried `docker logout`, then, `docker login `but still same error

**Comment by saikumaru** [comment]
Yes, if you are mentioning about logging on to dockerhub on CMD and on Desktop, yes, I did logout and login again and retried the steps, but still failed.

**Comment by vishnu-narayanan** [comment]
>@vishnu-narayanan yes, if you are mentioning about logging on to dockerhub on CMD and on Desktop, yes, I did logout and login again and retried the steps, but still failed.

Yes, login in using the `docker login` command. 

@saikumaru  Are you still having issues ?

**Comment by tuanhuydev** [comment]
I can confirm the issue still happen after logging in. @vishnu-narayanan

**Comment by vishnu-narayanan** [comment]
Could you please confirm if you can run the production `dokcer-compose` file ?
https://www.chatwoot.com/docs/self-hosted/deployment/docker

**Comment by teawithfruit** [comment]
I've got the same problem an macOS when trying to make the development environment run with this description.
https://www.chatwoot.com/docs/contributing-guide/environment-setup/docker

**Comment by jordan-brough** [comment]
Same here, also macOS and also following https://www.chatwoot.com/docs/contributing-guide/environment-setup/docker.

```
$ sw_vers
ProductName:	macOS
ProductVersion:	12.2.1
BuildVersion:	21D62

$ docker --version
Docker version 20.10.12, build e91ed57

$ docker-compose --version
Docker Compose version v2.2.3

$ git log -1 --oneline
a3cb26a3 (HEAD -> develop, origin/develop, origin/HEAD) fix: AudioContext warning when loading widget on Chrome (#3956)

$ docker-compose up --build
...
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0:
failed to create LLB definition: pull access denied, repository does not exist or may require authorization:
server message: insufficient_scope: authorization failed                                                                                        
```

Full output of `docker-compose`:
<details><summary>click here</summary>

```
docker-compose up --build
[+] Building 1.5s (14/42)                                                                                                                                                     
 => [chatwoot:development internal] load build definition from Dockerfile                                                                                                0.0s
 => => transferring dockerfile: 2.77kB                                                                                                                                   0.0s
 => [chatwoot-webpack:development internal] load build definition from webpack.Dockerfile                                                                                0.0s
 => => transferring dockerfile: 158B                                                                                                                                     0.0s
 => [chatwoot-rails:development internal] load build definition from Dockerfile                                                                                          0.0s
 => => transferring dockerfile: 2.77kB                                                                                                                                   0.0s
 => [chatwoot:development internal] load .dockerignore                                                                                                                   0.0s
 => => transferring context: 252B                                                                                                                                        0.0s
 => [chatwoot-webpack:development internal] load .dockerignore                                                                                                           0.0s
 => => transferring context: 252B                                                                                                                                        0.0s
 => [chatwoot-rails:development internal] load .dockerignore                                                                                                             0.0s
 => => transferring context: 252B                                                                                                                                        0.0s
 => [chatwoot-rails:development internal] load metadata for docker.io/library/ruby:3.0.2-alpine                                                                          0.0s
 => [chatwoot-rails:development pre-builder  1/13] FROM docker.io/library/ruby:3.0.2-alpine                                                                              0.1s
 => CANCELED [chatwoot:development internal] load build context                                                                                                          1.2s
 => => transferring context: 14.83MB                                                                                                                                     1.1s
 => ERROR [chatwoot-webpack:development internal] load metadata for docker.io/library/chatwoot:development                                                               1.1s
 => CANCELED [chatwoot-rails:development internal] load build context                                                                                                    1.1s
 => => transferring context: 10.62MB                                                                                                                                     1.0s
 => CANCELED [chatwoot-rails:development pre-builder  2/13] RUN apk update   && apk add     openssl     tar     build-base     tzdata     postgresql-dev     postgresql  1.2s
 => CANCELED [chatwoot-rails:development stage-1 2/6] RUN apk add --update --no-cache     openssl     tzdata     postgresql-client     imagemagick     git   && gem ins  1.2s
 => [auth] library/chatwoot:pull token for registry-1.docker.io                                                                                                          0.0s
------                                                                                                                                                                        
 > [chatwoot-webpack:development internal] load metadata for docker.io/library/chatwoot:development:                                                                          
------                                                                                                                                                                        
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed                                                                                        
```
</details>

**Comment by jordan-brough** [comment]
Where is `FROM chatwoot:development` supposed to come from? I don't see the `development` tag here:
https://hub.docker.com/r/chatwoot/chatwoot/tags

Is it supposed to be `chatwoot:develop`?

**Comment by jordan-brough** [comment]
### tl;dr: Increase the memory available to Docker in order to build the `grpc` gem.

[On macOS](https://docs.docker.com/desktop/mac/#resources), or [on Windows](https://docs.docker.com/desktop/windows/#resources).
Increasing `Memory` from 2GB to 4GB fixed the problem for me.

Alternatively: If you don't need `google-cloud-dialogflow` then remove [google-cloud-dialogflow from the Gemfile](https://github.com/chatwoot/chatwoot/blob/v2.2.1/Gemfile#L89-L90) and then run `bundle install`.  `grpc` is only currently included because of `google-cloud-dialogflow`.

---------------------

### More details
Looks like the above error is a due to the [base "chatwoot:development" image](https://github.com/chatwoot/chatwoot/blob/v2.2.1/docker-compose.yaml#L4-L16) failing during `bundle install`.

If I change [the bundle `-j` option](https://github.com/chatwoot/chatwoot/blob/v2.2.1/docker/Dockerfile#L45-L46) (parallel jobs) to `1` and run `docker build` directly I get this:
```
docker build . \
  --file docker/Dockerfile \
  --build-arg BUNDLE_WITHOUT='' \
  --build-arg EXECJS_RUNTIME='Node' \
  --build-arg RAILS_ENV='development' \
  --build-arg RAILS_SERVE_STATIC_FILES='false'

...
#13 183.4 Installing grpc 1.43.1 with native extensions
------
executor failed running [/bin/sh -c if [ "$RAILS_ENV" = "production" ]; then   bundle config set without 'development test'; bundle install -j 4 -r 3;   else bundle install -j 1 -r 3;   fi]:
exit code: 137
```
i.e. it fails with [error 137 (out of memory)](https://codefresh.io/docs/docs/troubleshooting/common-issues/error-code-137/) while compiling native extensions for `grpc`.
And `grpc` is included because of [google-cloud-dialogflow](https://github.com/chatwoot/chatwoot/blob/v2.2.1/Gemfile#L89-L90).

**Comment by sojan-official** [comment]
@jordan-brough Thanks a lot for reporting back

**Comment by sojan-official** [comment]
closing this based on the recommended solution. lets reopen if the issue persists

**Comment by kylemclaren** [comment]
For anyone running into this, if you're running `docker-compose`, try `docker compose`. And vice versa.

As per comment [here](https://github.com/docker/compose/issues/8449#issuecomment-914174592).
